### PR TITLE
Fix gocean alg psyir error (part of #753)

### LINKED ISSUE #753: [PSyIR] translate the algorithm layer to PSyIR
The algorithm layer is currently parsed and transformed with fparser2. We should use the PSyIR instead for flexibility and to support potential future optimisations. The idea would be to parse to low level PSyIR then to recognise invoke regions, Kernels and arguments with particular types (fields, scalars, ...) and have domain-specific nodes for these. In the first instance we would need to be able to re-create the existing algorithm transformation performed by fparser2.

Once the above is working then additional functionality to try to determine the datatypes, function spaces etc. of algorithm arguments (within invokes) that passed into the PSy-layer. This could then be used to perform static correctness checks between algorithm fields and kernel arguments (specified with metadata).

This would complement issue #79.

### PR DESCRIPTION
**Pull request by rupertford**
Fixes a bug reported by @hiker in the new psyir-based algorithm tranformation for the gocean1.0 api. With these changes the game-of-life example provided by @hiker compiles and runs.

### CODE REVIEW

**Comment by rupertford** [comment]
Changes were based on those suggested by @sergisiso (thanks).

**Comment by rupertford** [comment]
Ready for review from any of @arporter, @sergisiso, @hiker, @TeranIvy. This is a small bug fix.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for fixing this.
A few things need tidying up.
All tests and examples are OK with compilation.
This is a small bug fix so no docs need updating.
There is still a TODO referring to #753.
I've checked all files with pylint.

File: src/psyclone/domain/common/algorithm/psyir.py [hunk 1]
```diff
@@ -216,7 +217,9 @@ def create_psylayer_symbol_root_names(self):
             # we temporarily replicate this functionality. Eventually we
             # will merge. Note, a better future solution could be to use
             # the closest ancestor routine instead.
-            nodes = self.root.walk(Routine)
+            nodes = self.root.walk((Routine, Container))
+            nodes = [node for node in nodes if not
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/domain/common/algorithm/psyir.py:221]: Yes that is better thanks.

File: src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py [hunk 2]
```diff
@@ -283,8 +314,9 @@ def test_aic_defroutineroot_name_error():
 
 def test_aic_createpsylayersymbolrootnames():
     '''Check that the create_psylayer_symbol_root_names method behaves in
-    the expected way, i.e. creates and stores a root name for a
-    routine_symbol and a container symbol.
+    the expected way when the name comes from a subroutine, a module
+    and when it has a filecontainer, i.e. it creates and stores a root
+    name for a routine_symbol and a container symbol.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py:319]: Oops.

File: src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py [hunk 3]
```diff
@@ -295,23 +327,40 @@ def test_aic_createpsylayersymbolrootnames():
...
 
-    # Check that the names are only created once.
-    routine_root_name_tmp = invoke._psylayer_routine_root_name
-    container_root_name_tmp = invoke._psylayer_container_root_name
-    invoke.create_psylayer_symbol_root_names()
-    assert invoke._psylayer_routine_root_name is routine_root_name_tmp
-    assert invoke._psylayer_container_root_name is container_root_name_tmp
+    # File container and module
+    psyir = create_alg_psyir(code)
+    invoke = psyir.children[0].children[0][0]
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py:355]: In the next line we call `_check_alg_names()` and this already does that assert.

File: src/psyclone/domain/common/algorithm/psyir.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # -----------------------------------------------------------------------------
 # BSD 3-Clause License
 #
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/domain/common/algorithm/psyir.py:3]: The TODO is correct, it is the title of this PR that was wrong. This PR should not close #753 as there is more to do. I've changed the title.

**Comment by rupertford** [comment]
Ready for next review from @arporter

File: src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py:355]: OK.

File: src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py [hunk 5]
```diff
@@ -295,23 +327,40 @@ def test_aic_createpsylayersymbolrootnames():
...
+    _check_alg_names(invoke, "psy_my_mod")
+
+    # Module, no FileContainer
+    psyir = create_alg_psyir(code)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/algorithm/alg_psyir_test.py:359]: Oh yes, I see that `_check_alg_names` does more than I realised.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
All tests and examples are OK with compilation.
All docs build OK.

**Pull request merged**